src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -265,19 +255,34 @@ public void OnRemoved()
                 InMemoryStorage.DropCache(Analyzer);
             }
 
-            /// <summary>
-            /// Get the unique state name for the given analyzer.
-            /// Note that this name is used by the underlying persistence stream of the corresponding <see cref="ProjectState"/> to Read/Write diagnostic data into the stream.
-            /// If any two distinct analyzer have the same diagnostic state name, we will end up sharing the persistence stream between them, leading to duplicate/missing/incorrect diagnostic data.
-            /// </summary>
-            private static ValueTuple<string, VersionStamp> GetNameAndVersion(DiagnosticAnalyzer analyzer)
+            private sealed class AnalyzerTypeData
             {
-                Contract.ThrowIfNull(analyzer);
+                internal const string UserDiagnosticsPrefixTableName = "<UserDiagnostics2>";
+                private static readonly ConcurrentDictionary<Type, AnalyzerTypeData> s_analyzerTypeData
```

Comment:
if you are going to use concurrent dictionary, I think you can just pass in analyzerId? 

and then you dont need this.

Debug.Assert(_analyzerTypeData.StateName == $"{AnalyzerTypeData.UserDiagnosticsPrefixTableName}_{analyzerId}", "Expected persistence information for analyzer instance to be derived from its type alone.");
-

Reply:
also, concurrent dictionary we usually put concurrencylevel: 2. reason is paul herrington found long time ago it allocates 1 slot per new thread it got used and with away, it can be a lot.

Reply:
:arrow_right: I have most confidence that `Type` is unique here and will not conflict in error. The additional configuration shouldn't be needed because we are only creating one instance. I code where potentially thousands of instances are created, it's much more important.

Reply:
but, if Type is unique and analyzerId is not then Statename between 2 types are same and everything goes down from there. since it is supposed to be unique.

Reply:
I'm going to leave it like this for now since `Type` doesn't have any performance-related disadvantages, but if switching will result in simplifications of other areas of the code feel free to file an issue or create a follow-up PR. 😄

Reply:
sure, I am fine as well. I will do some clean up when I have chance to touch code around this.